src/devices/bus/isa/sblaster_lle.cpp

```diff
@@ -0,0 +1,415 @@
+
+void isa8_sblaster_2_0_lle_device::device_add_mconfig(machine_config &config)
+{
+	I80C51(config, m_dsp, XTAL_DSP); // XTL1
+	m_dsp->set_addrmap(AS_PROGRAM, &isa8_sblaster_2_0_lle_device::map_dsp_program);
+	m_dsp->set_addrmap(AS_IO, &isa8_sblaster_2_0_lle_device::map_dsp_io);
+
+	// No inputs on ports 0 or 1
+	m_dsp->port_in_cb<2>().set(FUNC(isa8_sblaster_2_0_lle_device::dsp_port2_r));
+	m_dsp->port_in_cb<3>().set(FUNC(isa8_sblaster_2_0_lle_device::dsp_port3_r));
+	// No outputs on port 0
+	m_dsp->port_out_cb<1>().set(m_dac, FUNC(dac_byte_interface::data_w));               // DSP port 1 is hooked right up to the DAC.
+	m_dsp->port_out_cb<2>().set(FUNC(isa8_sblaster_2_0_lle_device::dsp_port2_w));
+	m_dsp->port_out_cb<3>().set(FUNC(isa8_sblaster_2_0_lle_device::dsp_port3_w));
+
+	PC_JOY(config, m_joy);
+
+	SPEAKER(config, m_speaker).front_center();
+	MC1408(config, m_dac, 0).add_route(ALL_OUTPUTS, m_speaker, 0.5);
+
+	YM3812(config, m_ym3812, CLOCK_YM3812);
+	m_ym3812->add_route(ALL_OUTPUTS, m_speaker, 1.0);
+	/* no CMS */
+}
+
+static INPUT_PORTS_START( sb_dsw )
+	PORT_START("JP1")
+	PORT_CONFNAME( 0x03, 0x01, "I/O Base Address")
+	PORT_CONFSETTING(    0x01, "220h")
+	PORT_CONFSETTING(    0x02, "240h")
+
+	PORT_START("JP4")
+	PORT_CONFNAME( 0x0f, 0x01, "IRQ Line")
+	PORT_CONFSETTING(    0x08, "IRQ2/9" )
+	PORT_CONFSETTING(    0x04, "IRQ3" )
+	PORT_CONFSETTING(    0x02, "IRQ5")
+	PORT_CONFSETTING(    0x01, "IRQ7")
+INPUT_PORTS_END
+
+ioport_constructor isa8_sblaster_2_0_lle_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME( sb_dsw );
+}
+
+isa8_sblaster_2_0_lle_device::isa8_sblaster_2_0_lle_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	device_t(mconfig, ISA8_SOUND_BLASTER_2_0_LLE, tag, owner, clock),
+	device_isa8_card_interface(mconfig, *this),
+	m_dsp(*this, "dsp"),
+	m_joy(*this, "pc_joy"),
+	m_ym3812(*this, "ym3812"),
+	m_dac(*this, "dac"),
+	m_speaker(*this, "speaker"),
+	m_jp1(*this, "JP1"),
+	m_jp4(*this, "JP4")
+{
+}
+
+void isa8_sblaster_2_0_lle_device::device_start()
+{
+	m_installed_base_io = 0;
+	set_isa_device();
+}
+
+void isa8_sblaster_2_0_lle_device::device_reset()
+{
+	install_io();
+	lower_irq();
+	lower_dma();
+
+	// DMA is not configurable.
+	m_isa->set_dma_channel(1, this, false);
+
+	m_irq_in_flag = false;
+	m_irequest = true;
+	m_drequest = false;
+
+	dsp_reset_w(0, 0);
+
+	m_irq_raised = 0;
+	m_dma_raised = false;
+}
+
+void isa8_sblaster_2_0_lle_device::install_io()
+{
+	uint16_t base = (m_jp1->read() & 0x02) ? 0x240 : 0x220;
+
+	if(m_installed_base_io != 0)
+	{
+		// Unmap if already mapped.
+		m_isa->unmap_device(m_installed_base_io+0x6, m_installed_base_io+0x7);
+		m_isa->unmap_device(m_installed_base_io+0x8, m_installed_base_io+0x9);
+		m_isa->unmap_device(m_installed_base_io+0xa, m_installed_base_io+0xb);
+		m_isa->unmap_device(m_installed_base_io+0xc, m_installed_base_io+0xd);
+		m_isa->unmap_device(m_installed_base_io+0xe, m_installed_base_io+0xf);
+	}
```

Comment:
It isn’t safe to do this, because there’s nothing to stop some other device installing itself over the top of your handlers in the mean time, and this will unmap that as well.  The least bad thing to do is to only install handlers the first time, requiring a hard reset to change the base address, etc.